crates/goose/src/agent.rs

```diff
@@ -59,37 +96,136 @@ impl Agent {
 
     /// Dispatch a single tool call to the appropriate system
     async fn dispatch_tool_call(&self, tool_call: ToolCall) -> AgentResult<Value> {
-        let system = self.get_system_for_tool(&tool_call.name)
+        let system = self
+            .get_system_for_tool(&tool_call.name)
             .ok_or_else(|| AgentError::ToolNotFound(tool_call.name.clone()))?;
 
-        let tool_name = tool_call.name.split("__").nth(1)
+        let tool_name = tool_call
+            .name
+            .split("__")
+            .nth(1)
             .ok_or_else(|| AgentError::InvalidToolName(tool_call.name.clone()))?;
         let system_tool_call = ToolCall::new(tool_name, tool_call.parameters);
-        
+
         system.call(system_tool_call).await
     }
 
+    fn get_system_prompt(&self) -> AgentResult<String> {
+        let mut context = HashMap::new();
+        let systems_info: Vec<SystemInfo> = self.systems
+            .iter()
+            .map(|system| SystemInfo::new(
+                system.name(),
+                system.description(),
+                system.instructions(),
+            ))
+            .collect();
+
+        context.insert("systems", systems_info);
+        load_prompt_file("system.md", &context)
+            .map_err(|e| AgentError::Internal(e.to_string()))
+    }
+
+    /// Fetches the current status of all systems and formats it as a status message
+    async fn get_system_status(&self) -> AgentResult<String> {
+        // Get status of all systems
+        let status = if !self.systems.is_empty() {
+            let mut context = HashMap::new();
+            let mut systems_status: Vec<SystemStatus> = Vec::new();
+            for system in &self.systems {
+                let system_status = system
+                    .status()
+                    .await
+                    .map_err(|e| AgentError::Internal(e.to_string()))?;
+
+                // Format the status into a readable string
+                let status_str = serde_json::to_string(&system_status).unwrap_or_default();
+
+                systems_status.push(SystemStatus::new(
+                    system.name(),
+                    status_str,
+                ));
+            }
+            context.insert("systems", systems_status);
+            load_prompt_file("status.md", &context).map_err(|e| AgentError::Internal(e.to_string()))?
+        } else {
+            "No systems loaded".to_string()
+        };
+
+        Ok(status)
+    }
+
+    // Initialize a new reply round, which may call multiple tools
+    // NOTE this is a simple no-op in this implementation
+    // NOTE this is a potential home for summarization, checkpointing, planning
+    async fn rewrite_messages_loopstart(&self, messages: &mut Vec<Message>, status: String) -> AgentResult<()> {
+        // Create tool use message for status check
```

Comment:
nit: should we put the status request/response at the end of the messages?

Reply:
yeah that would make sense unless we need `cwd` for the first turn which i doubt?